### ISSUE TITLE
unreachable code is not detected if it depends on never type fallback

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/rust#146085]
I tried this code:

```rust
fn main() {
    (|| panic!())();
    ();
}
```

I expected to see this happen: `();` is marked as unreachable code

Instead, this happened: there is no warning.

This happens because we emit unreachable code lint in `rustc_hir_typeck`, while checking expressions (we track it via [`FnCtxt::diverges`](https://github.com/rust-lang/rust/blob/84a17470220e7adf249b18d7c0178dfbede89462/compiler/rustc_hir_typeck/src/fn_ctxt/mod.rs#L106)), which means that we can have unresolved inference variables.

In this case, when checking code reachibility we see that the type of ` (|| panic!())()` is a (diverging) inference variable and not `!` so we don't set `FnCtxt::diverges` to "always". Later this inference variable _is_ set to `!` (NB: currently only on edition = 2024), but at this point we don't check for unreachable code.

### Meta

`rustc --version --verbose`:
```
1.91.0-nightly (2025-08-31 07d246fc6dc227903da2)
```
